RFCs/FS-1152-For-loop-with-accumulation.md

```diff
@@ -0,0 +1,474 @@
+
+This RFC covers the detailed proposal for this suggestion.
+
+- [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/1362)
+- [ ] Approved in principle
+- [ ] [Implementation](https://github.com/dotnet/fsharp/pull/FILL-ME-IN)
+- [x] [Discussion](https://github.com/fsharp/fslang-design/discussions/808)
+
+# Summary
+
+A new syntax for `for`-loops that allow accumulation is introduced that is intuitive, teachable to new users, easily documentable and without too much unexpected behavior (all 4 phrases are just [Principle of Least Surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment)). It also derives and provides for useful semantics (a reasonably learned F# programmer will reliably reach for in practice, enabling the developers to make their lives easier). It is also rigorous (other features rely on this being sensible and future features may build on it).
+```fs
+for <pat> = <expr> with <pat> in <expr> do
+    <expr>
+```
+The value of the loop body is used to update the accumulator which is returned in the end.
+
+# Motivation
+
+Currently, to do this, we need an additional mutable variable. For example,
+```fs
+let mutable sentence = "" // mutable accumulator
+for word in ["Hello"; " "; "World"; "!"] do
+    sentence <- sentence + word
+printfn "%s" sentence
+```
+
+However, the mutable variable leaks outside the loop body, which is undesirable.
+
+Moreover, the accumulator must be mutable - which is against functional immutable semantics.
+The danger with mutable stateful objects is that it further encourages globally mutable state against functionally immutable design. This is why `fold` exists: to encapsulate mutability and temporary variables as well.
+
+It is proposed that the accumulator be embeddable into the loop itself, with a familiar and sensible parameter order the same as the fold lambda body, i.e. accumulator first, sequence second, such that refactors from `fold`s are easy:
+```fs
+for sentence = "" with word in ["Hello"; " "; "World"; "!"] do
+    sentence + word
+|> printfn "%s"
+```
+The return value of the loop body updates the accumulator. This is a synthesis of `fold`s with loops, therefore it can be called a "fold loop". There is no direct equivalent in other languages, they either have `fold` with a lambda or `for` loops with no accumulator.
+
+This offers a desirable middle ground between `for` loops that must only have side-effects (returning `unit`)
+and purely functional `fold`s that must be in lambda form. It is declarative which removes concerns about order of operations. It encourages functional pureness without the shortcomings of `fold`.
+
+`fold` is inferior to this syntax because it is hard to understand.
+```fs
+["Hello"; " "; "World"; "!"]
+|> Seq.fold (fun sentence -> sentence + word (*loop body but shown as a lambda!*)) "" // initial state is placed last??
```

Comment:
fold does not require placing the data last so this comment is inaccurate

Reply:
Updated it